### ISSUE TITLE
FEATURE REQUEST: define tasks as PHP variables so the logic can be included in a task override

### ISSUE THREAD

**Issue by arderyp** [deployphp/deployer#2123]
I'm guessing this idea has already been discussed, but I couldn't find an issue about it.

Right now tasks are defined as such:

```
desc('Rollback to previous release');
task('rollback', function () {
    $releases = get('releases_list');

    if (isset($releases[1])) {
        $releaseDir = "{{deploy_path}}/releases/{$releases[1]}";

        // Symlink to old release.
        run("cd {{deploy_path}} && {{bin/symlink}} $releaseDir current");

        // Remove release
        run("rm -rf {{deploy_path}}/releases/{$releases[0]}");


        writeln("<info>rollback</info> to {$releases[1]} release was <success>successful</success>");
    } else {
        throw new Exception("No more releases you can revert to.");
    }
});
```

lets say I want to override the rollback task and add a condition to be executed before rolling back, which would throw an exception in certain cases to prevent the rollback.  Right now, I'd have to do:

```
# My Code
after('deploy:failed',  'deploy:rollback');  // Throw exception in "deploy:rollback" to prevent falling into "rollback"
after('deploy:rollback','rollback');

task('deploy:rollback', function() {
    if ($myCondition) {
        throw new \Exception("woops");
    }
});
```

It would be handy, however, to have something like this:

```
# Deployer
$taskRollback = function() {
    $releases = get('releases_list');

    if (isset($releases[1])) {
        $releaseDir = "{{deploy_path}}/releases/{$releases[1]}";

        // Symlink to old release.
        run("cd {{deploy_path}} && {{bin/symlink}} $releaseDir current");

        // Remove release
        run("rm -rf {{deploy_path}}/releases/{$releases[0]}");

        writeln("<info>rollback</info> to {$releases[1]} release was <success>successful</success>");
    } else {
        throw new Exception("No more releases you can revert to.");
    }
};
task('rollback', $taskRollback);

# My code
task('rollback', function() {
    if ($myCondition) {
        throw new \Exception("woops");
    }
    $taskRollback();
});
```

This is basically allows the same functionality as extending a PHP class, overriding a method, and then calling `parent::method()` within it to decorate that same exact functionality.

I realize that using `after()` gets the job done in the above scenario, but deep customization can create a confusing nest of `before()` and `after()` calls that are hard to follow.  Furthermore, let's say we also wanted to trigger another set of core/recipe code in my overriden task--unlocking the deployment, for example:

```
# Deployer
$taskRollback = function() {
    $releases = get('releases_list');

    if (isset($releases[1])) {
        $releaseDir = "{{deploy_path}}/releases/{$releases[1]}";

        // Symlink to old release.
        run("cd {{deploy_path}} && {{bin/symlink}} $releaseDir current");

        // Remove release
        run("rm -rf {{deploy_path}}/releases/{$releases[0]}");

        writeln("<info>rollback</info> to {$releases[1]} release was <success>successful</success>");
    } else {
        throw new Exception("No more releases you can revert to.");
    }
};
$taskDeployUnlock = function () {
    run("rm -f {{deploy_path}}/.dep/deploy.lock");//always success
}
task('rollback', $taskRollback);
task('deploy:unlock', $taskDeployUnlock);

# My code
task('rollback', function() {
    if (my condition) {
        // Something went wrong, unlock the deployment and fail here
        $taskDeployUnlock();
        throw new \Exception("woops");
    }
    $taskRollback();
});
```

**Comment by arderyp** [comment]
sorry, I am only just now leanring of `invoke()`, nothing to see here :facepalm:

**Comment by arderyp** [comment]
Well, maybe this should stay open... I don't think invoke would work in the case of overriding a task.  Wouldn't the follwing lead to an infinite loop?

```
task("rollback", function () {
    if ($myCondition) {
        invoke("deploy:unlock");
        throw new \Exception("oops");
    }
    invoke("rollback");  // call parent rollback, infinite loop?
});
```

**Comment by antonmedv** [comment]
I’m thinking of adding $parent parametr.

**Comment by arderyp** [comment]
@antonmedv  interesting, do you care to explain more, I don't fully understand.

**Comment by antonmedv** [comment]
```diff
task("rollback", function ($parent) {
    if ($myCondition) {
        invoke("deploy:unlock");
        throw new \Exception("oops");
    }
+   $parent();
});
```

**Comment by antonmedv** [comment]
I am cleaning up on issues. Fill free to reopen if needed. 
Also, we have a new section available: [discussions](https://github.com/deployphp/deployer/discussions).
